test/spec/trackingModules/canvas/PopupMenuTracking.spec.js

```diff
@@ -0,0 +1,141 @@
+
+  const diagram = require('test/spec/simple.bpmn').default;
+  let bpmnJSTracking;
+
+  beforeEach(bootstrapModeler(diagram, {
+    additionalModules: [
+      BpmnJSTracking,
+      PopupMenuTracking
+    ]
+  }));
+
+  beforeEach(function() {
+    bpmnJSTracking = getBpmnJS().get('bpmnJSTracking');
+  });
+
+  beforeEach(inject(function(bpmnJSTracking) {
+    bpmnJSTracking.enable();
+  }));
+
+
+  describe('should subscribe', function() {
+
+    it('popupMenu.open', inject(function(eventBus) {
+
+      // given
+      const spy = sinon.spy(bpmnJSTracking, 'track');
+
+      // when
+      eventBus.fire('popupMenu.open');
+
+      // expect
+      expect(spy).to.have.been.calledOnce;
+    }));
+
+
+    it('popupMenu.trigger', inject(function(eventBus) {
+
+      // given
+      const spy = sinon.spy(bpmnJSTracking, 'track');
+
+      // when
+      eventBus.fire('popupMenu.trigger', { entry: { label: 'foo' }, event: {} });
+
+      // expect
+      expect(spy).to.have.been.calledOnce;
+    }));
+
+  });
+
+
+  describe('should track', function() {
+
+    it('popup menu open', inject(function(elementRegistry, selection) {
+
+      // given
+      const shape = elementRegistry.get('StartEvent_1');
+      selection.select(shape);
+
+      const spy = sinon.spy(function(event) {
+        const { name, data } = event;
+        expect(name).to.eql('canvas.popupMenu.open');
+        expect(data.selection).to.eql([ shape ]);
+      });
+
+      bpmnJSTracking.on('tracking.event', spy);
+
+      // when
+      triggerContextPad(shape, 'replace', 'click');
+
+      // then
+      expect(spy).to.have.been.calledOnce;
+
+    }));
+
+
+    it('popup menu trigger entry', inject(function(elementRegistry, selection, popupMenu) {
+
+      // given
+      const shape = elementRegistry.get('StartEvent_1');
+      selection.select(shape);
+
+      const spy = sinon.spy(function(event) {
+        if (event === 'canvas.popupMenu.trigger') {
+          const { data } = event;
+          expect(data.entryLabel).to.eql('Intermediate Throw Event');
+          expect(data.triggerType).to.eql('click');
+        }
+      });
+
+      bpmnJSTracking.on('tracking.event', spy);
+
+      // when
+      triggerContextPad(shape, 'replace', 'click');
+      triggerPopupMenu('replace-with-none-intermediate-throwing', 'click');
+
+      // then
+      expect(spy).to.have.been.called;
+      expect(spy.getCalls()[1].args[1].name).to.eql('canvas.popupMenu.trigger');
+    }));
+
+  });
+
+});
+
+
+// helpers /////////
+
+function triggerContextPad(shape, entry, action) {
+  const contextPad = getBpmnJS().get('contextPad');
+
+  const pad = contextPad.getPad(shape),
+        html = pad.html,
+        target = domQuery(`[data-action="${entry}"]`, html);
+
+
+  target.click();
+}
+
+function triggerPopupMenu(entryId, action) {
+  const popupMenu = getBpmnJS().get('popupMenu');
+
+  const entry = domQuery(`[data-id="${entryId}"]`, popupMenu._current.container);
```

Comment:
I'd rather not access private properties (popupMenu._current); let's use selector instead.

Reply:
We also use this in the core libraries and create/append anything. Should we not or should we only do it in core libraries?

Reply:
I think it may be acceptable in the library where the private property is defined, but in a 3rd module we should avoid it. Notice that in the first case the CI will break when the change is introduced while here we have to adjust stuff on dependency update.

Reply:
So it should not be like I upgrade from 12.2.0 to 12.2.3 and suddenly my test case breaks.